[FLINK-17169][table-blink] Refactor BaseRow to use RowKind instead of byte header

wuchong opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This is a pre-step of FLINK-16996. We refactor `BaseRow#get/setHeader` to `get/setRowKind` in this PR. And update all the existing code to send "insert/delete/update_before/update_after" messages instead of "+/-" messages. 

## Brief change log

- Update all the code to use RowKind instead of header. 
- String representation of BaseRow is `I(a, b, c)` instead of `(+|a, b, c)` now.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes, the `RowKind`
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

wuchong commented:
cc @godfreyhe

wuchong commented:
cc @godfreyhe

libenchao commented on flink-core/src/main/java/org/apache/flink/types/RowKind.java:
```diff
@@ -52,5 +52,26 @@
+	 *     <li>"D" represents {@link #DELETE}.</li>
+	 * </ul>
+	 */
+	public String shortString() {
```
What if we add a parameter for each enum, for example, INSERT("I"), UPDATE_BEFORE("UB"). Then we don't need this switch-case?

libenchao commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java:
```diff
@@ -129,13 +130,15 @@ public int getArity() {
 		// first nullBitsSizeInBytes byte is header.
-		return segments[0].get(offset);
+		byte header = segments[0].get(offset);
+		return RowKind.values()[header];
```
IMHO, using enum's values and ordinal is not very elegant.
How about adding a field explicitly whose meaning is byte encoding for each enum?

libenchao commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/WindowEmitStrategy.scala:
```diff
@@ -63,7 +63,7 @@ class WindowEmitStrategy(
   def produceUpdates: Boolean = {
     if (isEventTime) {
-      allowLateness > 0 || earlyFireDelayEnabled || lateFireDelayEnabled
+      earlyFireDelayEnabled || lateFireDelayEnabled
```
If we only set allowLateness, then it won't work?

wuchong replied:
Yes. I think this is a bug when we merging blink code. A window aggregation can produce updates only when early fire or late fire is enabled. `allowLateness` is just a hint to the statebackend for how long to keep the state when window is closed. If late data never trigger emission, the window still produce insert-only messages.

libenchao replied:
Then we should modify `WindowOperatorBuilder.withAllowedLateness()` too?

godfreyhe replied:
It's better we fix this in another issue if it's a bug

wuchong replied:
I will still keep this fix in this PR to make the whole tests to pass. But I will split it into another commit.

godfreyhe replied:
please add a test to verify this bug

wuchong commented on flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/harness/GroupAggregateHarnessTest.scala:
```diff
@@ -75,57 +75,57 @@ class GroupAggregateHarnessTest(mode: StateBackendMode) extends HarnessTestBase(
-    testHarness.processElement(new StreamRecord(binaryrow("aaa", 1L: JLong), 1))
-    expectedOutput.add(new StreamRecord(binaryrow("aaa", 1L: JLong), 1))
+    // insertion
+    testHarness.processElement(binaryRecord(INSERT,"aaa", 1L: JLong))
```
Yes. It's unnecessary, because this is a processing-time aggregation which doesn't depend on event-time.

wuchong commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/DeduplicateFunctionHelper.java:
```diff
@@ -35,25 +37,29 @@
 			}
 		}
+		// in order for better performance, we don't have state for LastRow
+		// if not generate UPDATE_BEFORE, thus, we can't produce INSERT messages for first row.
```
Upsert databases uses `UPSERT` statement (or called REPLACE statement) which will replace values if the key exists, otherwise, it will insert a new row. So it works if all rows are `UPDATE`. 

Currently, the upsert sinks doesn't distinguish `INSERT` and `UPDATE` messages, they will be executed in `UPSERT` statements.

godfreyhe replied:
does all database support this ? I know mysql supports "INSERT ON DUPLICATE KEY UPDATE"

wuchong replied:
I changed the logic to emit INSERT for first row with the cost of additional state accessing.

wuchong commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/JoinedRow.java:
```diff
@@ -17,14 +17,16 @@
 	private BaseRow row1;
 	private BaseRow row2;
-	private byte header;
+	private RowKind rowKind = RowKind.INSERT;
```
This keeps the same behavior as before, default as `INSERT`. And this also aligns the behavior of our new `GenericRowData`. If we don't have a default value, then the RowKind will be nullable, and code maybe error-prone.

wuchong commented:
Thanks @godfreyhe and @libenchao for the reviewing. I have addressed the comments. 

I would appreciate if @beyond1920 could help to review the TopN operator and tests change, because I changed the logic of TopN to always produce UPDATE_BEFORE and UPDATE_AFTER for primary key (partitionKey+rownum).

wuchong commented on flink-core/src/main/java/org/apache/flink/types/RowKind.java:
```diff
@@ -47,10 +47,69 @@
+	 * Creates a {@link RowKind} enum with the given short string and byte value representation of
+	 * the {@link RowKind}.
+	 */
+	RowKind(String shortString, byte value) {
```
cc @twalthr , could you have a look the changes for RowKind?

Using a byte value representation will be much faster than using enum ordinal during de/serialization. In my local benchmark, byte value is 24x fater than ordianl. The disadvantage is that the IDEA code completion show some verbose information. 

Benchmark code: https://github.com/wuchong/my-benchmark/blob/master/src/main/java/myflink/EnumBenchmark.java

Benchmark Result:

```
# Run complete. Total time: 00:03:35

Benchmark                   Mode  Cnt      Score      Error   Units
EnumBenchmark.testOrdinal  thrpt   20    876.048 ±   18.128  ops/ms
EnumBenchmark.testValue    thrpt   20  20827.764 ± 2084.072  ops/ms
```

![image](https://user-images.githubusercontent.com/5378924/79822112-ff662600-83c2-11ea-9fd6-9ba3fafba927.png)

twalthr replied:
Thanks for pinging me about this. Sounds good to me. Thanks for performing benchmarks. I added some comments.

wuchong commented on flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/DeduplicateFunctionHelper.java:
```diff
@@ -35,26 +35,31 @@
+		// check message should be insert only.
+		Preconditions.checkArgument(currentRow.getRowKind() == RowKind.INSERT);
+		BaseRow preRow = state.value();
+		state.update(currentRow);
```
After offline disscussion, I will add an optimization configuration to always send AFTER instead of INSERT for first row. This can avoid state accessing but is disabled by default.

beyond1920 commented:
LGTM

godfreyhe commented on flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/util/GenericRowRecordSortComparator.java:
```diff
@@ -43,10 +44,10 @@ public GenericRowRecordSortComparator(int sortKeyIdx, LogicalType sortKeyType) {
+		RowKind kind1 = row1.getRowKind();
+		RowKind kind2 = row2.getRowKind();
+		if (kind1 != kind2) {
+			return kind1.ordinal() - kind2.ordinal();
```
use byte value

godfreyhe commented:
Thanks for the update, LGTM

wuchong commented:
Thanks for the reviewing @beyond1920 @godfreyhe , will merge this once Travis is green.

twalthr commented on flink-core/src/main/java/org/apache/flink/types/RowKind.java:
```diff
@@ -47,10 +47,69 @@
+	/**
+	 * Returns a short string representation of this {@link RowKind}.
+	 *
+	 * <p><ul>
```
I would remove this list. It is redundant and defined in the enum value already. Same for the list below.

wuchong replied:
That's true. But if users are looking at the Javadoc website, the comment is the only thing he/she can see, not the code. That's why I put the list in the Javadoc.

twalthr replied:
But users that only look at JavaDocs are usually not the ones that should work with byte values for performance reasons :D

wuchong replied:
Yes. But end users may care about the short string representation?

twalthr commented on flink-core/src/main/java/org/apache/flink/types/RowKind.java:
```diff
@@ -47,10 +47,69 @@
+			case 2: return UPDATE_AFTER;
+			case 3: return DELETE;
+			default: throw new UnsupportedOperationException(
+				"Do not support to parse byte value " + value + " to RowKind.");
```
nit: `Unsupported byte value ' + value + ' for row kind.`

twalthr commented on flink-core/src/main/java/org/apache/flink/types/RowKind.java:
```diff
@@ -38,7 +38,7 @@
 	 * update of a row that is not uniquely identifiable by a key.
 	 */
-	UPDATE_BEFORE,
+	UPDATE_BEFORE("UB", (byte) 1),
```
One comment about `UB` and `UA`: Have you thought about just calling them `B` and `A`? I know this might also be confusing but having a single character would look nicer when printed to a console otherwise rows will be like:
```
I[]
UB[]
D[]
```

wuchong replied:
I thought about this. The reason I pick `UB` and `UA` is because Oracle uses `I` `D` `UO` `UN` to represent these 4 kinds. I think the "U" in the "UB" string is more meaningful than just a single "B".

Another idea is padding `+` and `-` to `I` `D` `U` to make them have the same length and still can be distinguished easily. What do you think?

```
I+ | a | 12 | FDF
U- | b | 13 | EFC
U+ | b | 14 | CDE
D- | d | 15 | IKH
```

twalthr replied:
Sounds good to me. It nicely shows whether they add or remove rows.

twalthr commented on flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/CalcITCase.scala:
```diff
@@ -66,7 +66,7 @@ class CalcITCase extends StreamingTestBase {
     env.execute()
 
-    val expected = List("I(1,1,1)")
+    val expected = List("I+(1,1,1)")
```
after looking at `I+(1,Alice)`  I’m a bit concerned because it looks like math
other way around `+I(1,Alice)` ?

wuchong replied:
```
+I | a | 12 | FDF
-U | b | 13 | EFC
+U | b | 14 | CDE
-D | d | 15 | IKH
```

wuchong commented:
Hi Timo, I will merge this PR now to unblock following works. We can continue to discuss the Javadoc or string representation and update later if needed.
